### ISSUE TITLE
Parallel WaveNet

### ISSUE THREAD

**Issue by DabiaoMa** [r9y9/wavenet_vocoder#7]
Hi,

I am currently working on parallel WaveNet and noticed that it is one of your planned TODOS. 

Till now I can only generate wavs that are distinguishable but with a lot of noise. From my experience, the power loss is really crucial, and we need to do sufficient sampling per step to approximately compute the cross entropy term(If not sufficient, the high variance of gradients probably makes the model collapse).

Do you have anything in mind that may help? Thanks

**Comment by r9y9** [comment]
Sorry, I haven't really look into Parallel WaveNet in details yet. That's on my TODO list but lower priority, since I want to see quality improvement by WaveNet first. Just curious, did you get human-like speech quality by teacher WaveNet?

**Comment by DabiaoMa** [comment]
Well I can tell what the generated wav is speaking, but the overall audio is of very bad quality. 
The distilling from teacher WaveNet really works though some details are still to be corrected.

**Comment by jamestang0219** [comment]
@DabiaoMa You have already get teacher model of WaveNet? Could you upload some samples which generate from teacher model, thank you!

**Comment by npuichigo** [comment]
@DabiaoMa Can you share your parameters of your teacher model, for example, the minimum value of the log scale of mixture of logistic distributions? Because I find that this parameter is highly related to audio quality. As to the student network, I also find that the high variance of gradients make the log scale of student network overflow easily, and it's hard to get distinguishable waves.

**Comment by DabiaoMa** [comment]
@npuichigo I got a loss as low as about 1.6 when training Teacher model(a mixture of 10 logistic dist.) and the synthesized wavs are quite good(with real acoustic features, not with generated ones from a parametric model). To train the student network, I sampled 5 times for one local acoustic condition to make the gradient stable but failed...

**Comment by npuichigo** [comment]
@DabiaoMa What about your minimum value of the log scale? I think -7 is not enough to generate good samples, maybe with little noise.

**Comment by DabiaoMa** [comment]
@npuichigo  I am using similar implementation with the one used in https://github.com/openai/pixel-cnn/blob/master/pixel_cnn_pp/nn.py. 
So yes I am using -7 as the minimum value of the log scale
But I do not use the extreme case when the class label is 0 or 255.

**Comment by npuichigo** [comment]
@DabiaoMa Thanks! By the way,  when training student network, have you encountered such a problem that the entropy of student network (lns + 2) often increases suddenly? I think sampling multiple times may  relieve this problem, but not quite ideal.

**Comment by DabiaoMa** [comment]
@npuichigo I have got a much better distilled student model last week that can produce wavs with less noise (within 1 epoch), The key point to avoid the student entropy to increase too much is the power loss term. The student entropy stays at the value about (3 + 2) in my case.

**Comment by npuichigo** [comment]
@DabiaoMa Thanks! I also found the power loss critical. But in my training process, I usually get large power loss (from thousands to tens of thousands), can you share the frame length/frame shift used in your stft calculation.

**Comment by hicolin** [comment]
@DabiaoMa Hi, could you describe how you implement the cross entropy term in kl divergence? in my implementation of the student wavenet, the kld loss seems not to contribute to the model training, which is pretty weird. Thanks a lot!

**Comment by npuichigo** [comment]
@hicolin I think it is important to sampling sufficient samples from the student distribution to evaluate the cross entropy term validly.

**Comment by hicolin** [comment]
@npuichigo  well, how many samples do you sample from the student distribution in one step or mini-batch?  To many samples are a little time-consuming for training after all.

**Comment by zhf459** [comment]
@npuichigo hi, what's the meaning of the description of  working out the  kld loss?
![question](https://user-images.githubusercontent.com/8691353/36885634-2a18642e-1e23-11e8-9851-2cc2e1583af7.jpg)

**Comment by npuichigo** [comment]
@zhf459 The H(Ps) term of kld can be solved analytically, while the H(Ps, Pt) can only be estimated by sampling, which I mean is that we first draw multiple different samples from the student and then use the teacher to score them. It's quite similar to the form of the reconstruction error in VAE.

**Comment by zhf459** [comment]
@npuichigo How to score them?Say I sample from student(1-mixture) to get X and I know the distribution of the teacher(10-mixture), so I have to get the log_Pt(by discretized_mix_logistic_loss) and Ps(sigmoid((x-mu)/s)) for H(Ps,Pt) =-sum(log_Pt*Ps) term, and H(Ps)=sum(log(s)+2), then kl_loss = H(Ps,Pt) - H(Ps) and average the this kl_loss for multi-samples? I'm not sure I understand correctly.

**Comment by npuichigo** [comment]
@zhf459 Since discretized mixture of logistic distribution is represented as a set of pi, u and s, I think you cannot use H(Ps,Pt) =-sum(log_Pt*Ps) like softmax. In fact, H(Ps, Pt)=E(-log(Pt(x))) where x ~ Ps, so you only need to feed the sampled x into the teacher to estimate the term, like what you do when training the teacher. The teacher loss is just the negative log probability.

**Comment by Beronx86** [comment]
@zhf459 Hi, iaf stack in parallel wavenet should output 3 parameters, `x`, `m` and `s`.  `x` is the predicted sample. `m` and `s` are the output distribution mean and scale. `x` can be calculated from `m`, `s` and the `random_input`. `x = m + random_input * s`.  According to the paper, `H_Ps` is to sum up `ln(s) + 2` according to the time dimension, assuming `s` is of shape `batch_size * time_steps`. When calculating `H_Ps_Pt`, we need sample from the student distribution for each time step (`Ps(x_t | x_<t)`) first. Since we can got `m` and `s` of the student distribution, we can do this. (refer to [logistic distribution](https://en.wikipedia.org/wiki/Logistic_distribution) for sampling from logistic distribution.)  Input the student's prediction `x` to the teacher, we got `Pt(x_t | x_<t)` described by mixture of logistic distribution parameters. Then we can input the student's sample to the teacher wavenet as the teacher's target to  evaluate the probability of the student's sample under the teacher's distribution, i.e. we get the value of `Pt(x_t | x_<t)`. Then average it by number of student's random samples, we get  `-H(Ps, Pt)`.

**Comment by zhf459** [comment]
@npuichigo  @Beronx86  Thank you all ,I will check it today.

**Comment by neverjoe** [comment]
some one can share your Parallel WaveNet generated wav?@npuichigo @Beronx86 @DabiaoMa

**Comment by Beronx86** [comment]
@neverjoe the result is not good enough. the predicted wave is very noisy.

**Comment by npuichigo** [comment]
@Beronx86 Have you ever seen anything that the entropy of student network suddenly changes? Can you share your student entropy after being stable?

**Comment by Beronx86** [comment]
The loss curve is also very noisy. still debugging.
<img width="364" alt="dingtalk20180306161359" src="https://user-images.githubusercontent.com/9522624/37021320-193d2074-215a-11e8-96b4-56e8c20ab096.png">
<img width="384" alt="dingtalk20180306161316" src="https://user-images.githubusercontent.com/9522624/37021321-196db0c2-215a-11e8-86ac-f897f4add586.png">
<img width="374" alt="dingtalk20180306161227" src="https://user-images.githubusercontent.com/9522624/37021322-19a4c2b0-215a-11e8-93de-4f19461efa4b.png">

**Comment by zhf459** [comment]
@Beronx86 have you tried only use the kl_loss as the loss of Parallel Wavenet?

**Comment by Beronx86** [comment]
Hi @zhf459 the previous loss figure is only using kl loss.

**Comment by npuichigo** [comment]
@Beronx86 What about the mean loss averaged over samples? Perhaps you sum over them?

**Comment by Beronx86** [comment]
@npuichigo sum over time steps.

**Comment by xuerq** [comment]
Hi

**Comment by xuerq** [comment]
training a IAF student modle bease on softmax teacher WaveNet seems feasible，but quite noisy 
 
from top to bottom：ground truth、teacher model synthesis、white noise and student model synthesis
![image](https://user-images.githubusercontent.com/16162414/37336692-22fe804c-26ed-11e8-855d-ae88905dba76.png)
![image](https://user-images.githubusercontent.com/16162414/37336724-34571994-26ed-11e8-9dca-741585ff909f.png)

**Comment by npuichigo** [comment]
@xuerq Hi, can you get noiseless results from a student model trained with a logistic distribution teacher?

**Comment by xuerq** [comment]
Hi @npuichigo , training a MoL WaveNet is very time-consuming and less stability than softmax from my experience . so I didn't try it

**Comment by neverjoe** [comment]
@xuerq great job, can u share your network or training params？

**Comment by xuerq** [comment]
@neverjoe I'm not sure whether my parallel wavenet implementation is correct or not, still working on it, but the teacher model is similar to deepmind's paper except for the ouput layer and the dilated convolution filter size (not 3)

**Comment by neverjoe** [comment]
someone can tell me how to sample from student model? I can't understand what's the input of teacher 
model? It's z*scale + mu, z is the student's input? Or np.random.logistic(mu, scale)?
@DabiaoMa @npuichigo @xuerq

**Comment by r9y9** [comment]
Please respect everyone's time and do not request too much with @ mentions.

**Comment by Beronx86** [comment]
@neverjoe the teacher's input is the student's prediction, so z is the student's random input. z * scale_tot + mean_tot is the student's prediction.

**Comment by npuichigo** [comment]
@Beronx86 So what about sampling mutiple times? Just sample multiple z to get multiple x, or fix z to predict scale_tot and mean_tot then sample multiple x from Logistic(scale_tot, mean_tot) which is epsilon * scale_tot + mean_tot (epsilon ~ Logistic(0, 1)).

**Comment by Beronx86** [comment]
I think the latter is correct.

**Comment by xuerq** [comment]
@neverjoe I agree with Beronx86

**Comment by jiqizaisikao** [comment]
Temp method that loss=H(ps,pt)-H(ps) sample from the student is time consuming,I have tried sampling from the teacher and  loss=H(pt,ps)-H(pt) but it seems that it will easily got nan value.Anybody know the reason?

**Comment by jiqizaisikao** [comment]
@Beronx86 Hi, how many samples do you used sampling from student?I tried 15 samples but i get nan value,wavenet  quantization is 256

**Comment by begeekmyfriend** [comment]
Here is another version for parallel https://github.com/andabi/parallel-wavenet-vocoder

**Comment by bfs18** [comment]
I implemented a parallel wavenet based on nsynth and got some positive results. The generated wave is noisy but intelligible. I am still working on it.
https://github.com/bfs18/nsynth_wavenet/blob/master/tests/pred_data-pwn-failed_cases/gen_LJ001-0001-stft_abs.wav

**Comment by zhang-jian** [comment]
Updated Parallel Wavenet paper http://proceedings.mlr.press/v80/oord18a/oord18a.pdf

**Comment by r9y9** [comment]
https://github.com/azraelkuan/parallel_wavenet_vocoder

Nice fork :+1:

**Comment by PengjuYan** [comment]
As for @npuichigo 's question about sampling multiple `x_t` from the student distribution, I don't think there's any difference between the two processes. According to (3) from the paper, the student distribution depends only on, and can be determined only by, the noise distribution and the total scale/shift `s_i`/`mu_i`. We know that both `s_i` and `mu_i` can be got by forwarding the normalizing flows from the inputs `z_{1:t-1}`. Therefore, calculating `x_t` by sampling `z_t` (the first method in the question) and sampling directly from resultant distribution `Logistic(mu_i, s_i)` (the second method) are exactly the same process.

**Comment by PengjuYan** [comment]
@xuerq I'm wondering how you train the IAF model if you use a non-differentiable categorical distribution (w.r.t. `x_t`)?

![screen shot 2018-09-06 at 17 44 06](https://user-images.githubusercontent.com/13349414/45149581-7ce41f00-b1fc-11e8-9a84-a82ef5490725.png)

**Comment by hxs7709** [comment]
Can anybody share a synthesised wav and the commands of training from para wavenet, such as this tree https://github.com/azraelkuan/parallel_wavenet_vocoder? 
In my test with Chinese, the wav from student model (trained for more than 1000k steps) is still very bad (the wav from teacher model is already good ), and I cannot see any improvement after 170k steps. 
Thank you.

**Comment by MlWoo** [comment]
I have implemented the clarinet vocoder which is parallel wavenet of Gaussian output . [My results](https://twitter.com/menglin_wu/status/1074408876030783488) is free of noise in silence area in a nutshell. Howerver, you could see the slight noise at start that shows it seems like a slow-to-warm-up network. I am figuring out what happend at start of the wav.

**Comment by MlWoo** [comment]
This is my result. ![](https://pbs.twimg.com/media/DukRKUIXQAAMeAI.jpg:large)

**Comment by dukanov** [comment]
@MlWoo could you provide samples to listen?

**Comment by MlWoo** [comment]
@dukanov ISorry to tell you that the model is trained only with Chinese mandarin. Results of English will be available in next week.

**Comment by MlWoo** [comment]
@dukanov Here is a [sample](https://twitter.com/menglin_wu/status/1076666568954822657) which comes from a early-stage clarinet vocoder model only trained 90k steps with 2 batch size.

**Comment by dukanov** [comment]
@MlWoo great! thanks!

Does presented low noise appears in the following training steps?

**Comment by MlWoo** [comment]
@dukanov  Talking is cheep, show [my sound.](https://twitter.com/menglin_wu/status/1077537730186207232)

**Comment by r9y9** [comment]
I'm afraid that I won't support parallel wavenet. https://github.com/ksw0306/ClariNet might be a good start to play with parallel wavenet if anyone interested.
